### PR TITLE
Remove prompts dependency

### PR DESCRIPTION
**Pull request by dgarcia360**
Closes https://github.com/adamtheturtle/sphinx-substitution-extensions/issues/166

### CODE REVIEW

**Comment by dgarcia360** [comment]
Thanks @adamtheturtle for creating this extension! I've been using it for a while, and it works like a charm. However, now I'm looking to remove the ``sphinx-prompts`` dependency because one of the projects I'm working on does not use it.

This pull request removes the dependency and updates the docs. I'm not sure how to add a test for this change because the user would always have the dependency installed if they load the developer dependencies defined in ``requirements/dev-requirements.txt``.  I appreciate any guidance you could give me to merge the PR. Many thanks!

File: src/sphinx_substitution_extensions/utils.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+def exists_dependency(
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/utils.py:1]: This change means that someone can do:

```python
from sphinx_substitution_extensions.utils import exists_dependency
```

Which adds to what looks like a public interface.
Therefore, I would prefer to use `_utils.py`.

However:

* I am not fond of the name `utils` at all. I would prefer to have a file name which refers only to this function than to create `utils` (or `common` or `general` or anything similar).
* I don't think that it is necessary to have this function here - it can be a private function in `__init__.py`.

File: sample/source/conf.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 """
 Sample ``conf.py``.
 """
+import os
```
Comment (adamtheturtle) [inline sample/source/conf.py:4]: * What does this change achieve?
  * Is that relevant to this PR?

If it is required and relevant:

* Can a comment be added so that maintainers understand why this is here?
* Can this be done with [`Pathlib`](https://docs.python.org/3/library/pathlib.html)?

File: src/sphinx_substitution_extensions/__init__.py [hunk 3]
```diff
@@ -27,15 +30,16 @@
...
-_EXISTING_PROMPT_DIRECTIVE: Directive = _EXISTING_DIRECTIVES['prompt']
-
+_EXISTING_PROMPT_DIRECTIVE = Directive
+if exists_dependency('sphinx-prompt'):
+    if 'prompt' not in _EXISTING_DIRECTIVES:
+        MESSAGE = (
+            'sphinx-prompt must be in the conf.py extensions list before '
+            'sphinx_substitution_extensions'
+        )
+        LOGGER.error(MESSAGE)
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py:40]: I think that this is somewhat strange behaviour:

* I create a project which is not using `sphinx-substitution-extensions` and not `sphinx-prompt`.
* I `pip install sphinx-prompt`.
* I now get an error in my logs.

Is that right? If so, it might be confusing, for example, if a user has multiple Sphinx projects within one Python project.

What do you think?

Maybe this should become a `LOGGER.warn`?

File: src/sphinx_substitution_extensions/__init__.py [hunk 4]
```diff
@@ -27,15 +30,16 @@
...
-if 'prompt' not in _EXISTING_DIRECTIVES:
-    MESSAGE = (
-        'sphinx-prompt must be in the conf.py extensions list before '
-        'sphinx_substitution_extensions'
-    )
-    LOGGER.error(MESSAGE)
-
-_EXISTING_PROMPT_DIRECTIVE: Directive = _EXISTING_DIRECTIVES['prompt']
-
+_EXISTING_PROMPT_DIRECTIVE = Directive
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py:33]: Why have this? Is it so that we do not error when subclassing `_EXISTING_PROMPT_DIRECTIVE` for `SubstitutionPrompt`?

If so, I wonder, can we do better?
In particular, it is a bit confusing that we have a variable `_EXISTING_PROMPT_DIRECTIVE`, which is very much not an existing prompt directive.

I also expect that `mypy` will have problems with this but I cannot test that now as it appears that PyPI is having issues.

For example:

* Adding a comment explaining this line (bare minimum), or preferably
* Only define `SubstitutionCodeBlock` if we have `sphinx-prompt`

File: src/sphinx_substitution_extensions/__init__.py [hunk 5]
```diff
@@ -27,15 +30,16 @@
...
+if exists_dependency('sphinx-prompt'):
+    if 'prompt' not in _EXISTING_DIRECTIVES:
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/__init__.py:35]: Personal preference / nit: Can we swap the order so we have `if 'prompt' in _EXISTING_DIRECTIVES` ?

File: requirements/dev-requirements.txt [hunk 6]
```diff
@@ -14,5 +14,6 @@ pylint==2.12.2
 pyroma==3.2
 pytest-cov==3.0.0  # Measure code coverage
 pytest==6.2.5
+sphinx-prompt>=0.1
```
Comment (adamtheturtle) [inline requirements/dev-requirements.txt:17]: I think that this is best as a requirement in a `prompt` (install extra), so our users can `pip install sphinx-subsititution-extensions[prompt]`.

I'm happy to discuss this.

**Review by adamtheturtle — CHANGES_REQUESTED** [review]
Thank you @dgarcia360 for your work and for your kind words. I very much appreciate the submission, and am happy to help get this to a mergeable state.

I have left some initial comments and questions.

As for testing, perhaps take a look at the existing GitHub Workflow for CI at https://github.com/adamtheturtle/sphinx-substitution-extensions/blob/master/.github/workflows/ci.yml.

Perhaps you could add a step after running tests which `pip uninstall`s `sphinx-prompt` and builds a sample project. I guess that would fail without this change but would succeed with this change.

100% coverage is required by CI. If that is very hard to reach, we can consider using `# pragma: no cover` with a clear explanation of why we are ignoring some code paths.

**Comment by dgarcia360** [comment]
Thanks @adamtheturtle for reviewing! I'm changing this PR to draft mode while I work on the reamining comments.  I'll ping you once I finish 👍

File: requirements/dev-requirements.txt (hunk 6 above)
Comment (dgarcia360) [reply requirements/dev-requirements.txt]: Sounds good! Today I learned about python extras 😄

File: sample/source/conf.py (hunk 2 above)
Comment (dgarcia360) [reply sample/source/conf.py]: I've deleted them - it's not relevant to this PR.

I was trying to build the sample docs with the command ``sphinx-build``, but got an error without those lines. So now I've tried building the docs with the command ``make build-sample`` and it seems to work as expected without those lines.

File: src/sphinx_substitution_extensions/__init__.py (hunk 3 above)
Comment (dgarcia360) [reply src/sphinx_substitution_extensions/__init__.py]: That's right, since you have the extension installed but you haven't added it yet to ``conf.py``.

I like the idea of changing it to a warning in case they have the dependency installed in case they share the same env for multiple Sphinx projects.

File: src/sphinx_substitution_extensions/__init__.py (hunk 4 above)
Comment (dgarcia360) [reply src/sphinx_substitution_extensions/__init__.py]: To not add the comment, I've done the following:

- Moved ``SubstitutionPrompt`` into a separate file ``extras.py``.
- Load the directive conditionally, only if ``sphinx-prompt`` is installed.
- Created ``shared.py`` for variables and functions that are common for both directives.

What do you think?

**Comment by dgarcia360** [comment]
@adamtheturtle Ready for review - with 100% coverage :) 

I added the following annotation to run some tests conditionally only when `sphinx-prompt` is installed:

```
_EXISTS_PROMPT_EXTENSION = _exists_dependency('sphinx-prompt')
_REASON = 'requires sphinx-prompt to be installed'

@pytest.mark.skipif(
    not _EXISTS_PROMPT_EXTENSION,
    reason=_REASON,
)
def test_prompt_specified_late(tmp_path: Path) -> None:
...
```

Please let me know if you would like to change anything else. Thanks!

File: src/sphinx_substitution_extensions/shared.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+from typing import Dict
+
+from docutils.parsers.rst import directives  # type: ignore
+from docutils.parsers.rst import Directive
+from sphinx.directives.code import CodeBlock
+
+_EXISTING_DIRS = directives._directives  # pylint: disable=protected-access
```
Comment (adamtheturtle) [inline src/sphinx_substitution_extensions/shared.py:7]: These variables are imported from elsewhere but really should not be as they are marked with underscores.
Instead - change the filename to start with an underscore and leave these without underscores.

File: tests/test_substitution_extensions_extras.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+from pathlib import Path
+from textwrap import dedent
+
+import pytest
+
+from sphinx_substitution_extensions import _exists_dependency
+
+_EXISTS_PROMPT_EXTENSION = _exists_dependency('sphinx-prompt')
+_REASON = 'requires sphinx-prompt to be installed'
+
```
Comment (adamtheturtle) [inline tests/test_substitution_extensions_extras.py:12]: Optional - use:

```python
pytestmark = [pytest.mark.skipif(not _EXISTS_PROMPT_EXTENSION, reason=_REASON)]
```

That way all the tests in the file get the `skipif` functionality, including new tests.

**Comment by adamtheturtle** [comment]
Thanks @dgarcia360 .
Please can you look into the build failures - they seem relevant.

**Comment by dgarcia360** [comment]
Thanks @adamtheturtle,

- After the latest changes, `ci.yml` is running without errors now.
-  `ci-versions.yml` still raises the following error ``Class cannot subclass "CodeBlock" (has type "Any")``. Let's see if I can type it. Otherwise, I was thinking about adding an annotation to silence the warning.

**Comment by dgarcia360** [comment]
Hi @adamtheturtle,

I've been trying different alternatives, but I always get to the same situation with ``mypy``:

``ci-versions.yml`` (Sphinx 2.0) raises: ``Class cannot subclass "CodeBlock" (has type "Any")``. However, if I add a ``type: ignore comment``, ``ci-versions.yml`` passes but ``ci.yaml`` (Sphinx latest) raises the error ``type: ignore unused``.

Related issue: https://github.com/python/mypy/issues/8823

I appreciate any tips you could provide to merge the PR.

**Comment by adamtheturtle** [comment]
@dgarcia360 Thanks for working on this.

Am I right in thinking that the problem is that the type check doesn't work with an older version of a dependency (either Sphinx or sphinx-prompt)?

If so, this is a good enough reason to bump the minimum version of that dependency.

You can:

* Bump it in the requirements file
* Bump it in the `ci-versions` workflow
* Mention the new requirement in the CHANGELOG

**Comment by dgarcia360** [comment]
Thanks for the tip! I've upgraded the minimum version and the warning disappeared, but now it raises others  (only with Sphinx 3.x):

```
src\sphinx_substitution_extensions\shared.py:6: error: Unused "type: ignore" comment
src\sphinx_substitution_extensions\extras.py:6: error: Unused "type: ignore" comment
src\sphinx_substitution_extensions\__init__.py:13: error: Unused "type: ignore" comment
Found 3 errors in 3 files (checked 8 source files)
```

If I remove the comments, ``mypi`` raises the warnings they are silencing with Sphinx 4.x. I couldn't find either how to silence the "unused comment" warning at line level.

To remove them, I opted for disabling ``warn_unused_ignores`` directly in ``pyproject.toml``. Most likely this issue will be solved if ``typeshed`` adds the type for Sphinx 4.x.

```
# See https://github.com/python/typeshed/issues/5755
warn_unused_ignores = false
```

Would you accept this solution? Now, both CI scripts seems to pass without errors.

**Comment by adamtheturtle** [comment]
Hi @dgarcia360 , thank you for your persistence.

I don't really understand the problem, and cannot so far reproduce `mypy` errors locally. If you can - can you please explain how?

Otherwise, it sounds like bumping the minimum Sphinx version to ~4 would solve the issue?

**Comment by dgarcia360** [comment]
@adamtheturtle Thanks for your time too!

> I don't really understand the problem, and cannot so far reproduce mypy errors locally. If you can - can you please explain how?

1. Set ``warn_unused_ignores = true`` in ``pyproject.toml``
2. Run ``pip install Sphinx==3.0``
3. Run ``make lint``. The console reports:
```
src\sphinx_substitution_extensions\shared.py:6: error: Unused "type: ignore" comment
src\sphinx_substitution_extensions\extras.py:6: error: Unused "type: ignore" comment
src\sphinx_substitution_extensions\__init__.py:13: error: Unused "type: ignore" comment
Found 3 errors in 3 files (checked 8 source files)
```

> Otherwise, it sounds like bumping the minimum Sphinx version to ~4 would solve the issue?

Yes, bumping the minimum requirement to Sphinx 4.0 solves the issue. However, I think it's a pity to lose Sphinx 3.0.0 compatibility since these ``type: ignore`` comments were added to solve an warning that only seems to appear since Sphinx 4.0 (https://github.com/python/typeshed/issues/5755). 

If we were to remove the ``type: ignore`` comments, Sphinx 3.0.0 would work without errors but Sphinx 4.0.0 would raise the 
 warning ``docutils.parsers.rst" has no attribute "directives"; maybe "Directive"?``. For this reason, I thought it would be best to not check for ``Unused "type: ignore" comment``, at least until https://github.com/python/typeshed/issues/5755 is solved.

**Comment by adamtheturtle** [comment]
Thank you @dgarcia360 .

Why is is bad to remove Sphinx 3 compatibility?
Do any users of this project benefit from Sphinx 3 compatibility?
They already have the solution of pinning to the now-current version of this package.

Personally, I care more about having `mypy` on the strictest settings.

**Comment by dgarcia360** [comment]
> They already have the solution of pinning to the now-current version of this package.
> Personally, I care more about having mypy on the strictest settings.

Fair enough, I've updated the version to Sphinx 4.0. Please let me know if you see anything else. Thanks!

**Review by adamtheturtle — APPROVED** [review]

**Comment by adamtheturtle** [comment]
Nice work!

**Comment by dgarcia360** [comment]
@adamtheturtle Would be possible to release a new version of the package in PyPI? Thanks!

**Comment by adamtheturtle** [comment]
See https://pypi.org/project/Sphinx-Substitution-Extensions/2022.2.16/.

**Comment by adamtheturtle** [comment]
I decided to remove this in a recent change because it was causing headaches. In particular, doing work on import (to figure out whether to warn) and checking whether the dependency existed meant that tests which ran multiple Sphinx builds conflicted with each other.

> However, now I'm looking to remove the sphinx-prompts dependency because one of the projects I'm working on does not use it.

I should have dug into this at the time - "so what?" is what I should have asked.
